File: go/voice/group_service.go
// SPDX-License-Identifier: MIT

package voice

import (
	"bytes"
	"context"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/bhengubv/aether-protocol/go/constants"
	"github.com/bhengubv/aether-protocol/go/protocol"
	"github.com/bhengubv/aether-protocol/go/routing"
)

// ──────────────────────────────────────────────
// JSON signaling types
// ──────────────────────────────────────────────

// GroupVoiceSignalingMessage is the JSON-encoded payload for group voice
// signaling packets (VoiceSignaling with group semantics).
type GroupVoiceSignalingMessage struct {
	Kind          string   `json:"kind"`
	CallID        string   `json:"call_id"`
	FromUhid      string   `json:"from_uhid"`
	ToUhid        string   `json:"to_uhid"`
	InvitedUhids  []string `json:"invited_uhids,omitempty"`
	KickedUhid    string   `json:"kicked_uhid,omitempty"`
	KeyGeneration uint32   `json:"key_generation,omitempty"`
}

// ──────────────────────────────────────────────
// Session state
// ──────────────────────────────────────────────

// GroupVoiceCallSession holds per-call runtime state for a group call.
type GroupVoiceCallSession struct {
	CallID        uuid.UUID
	HostUhid      string
	Members       map[string]struct{} // uhid -> present
	State         VoiceCallState
	KeyGeneration uint32
	Sequence      uint32
	mu            sync.Mutex
	CreatedAt     time.Time
}

// ──────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────

// GroupVoiceCallService manages group voice calls over the mesh.
// Up to constants.MaxGroupVoiceMembers participants are supported.
type GroupVoiceCallService struct {
	sender    routing.MeshSender
	localUhid string
	calls     sync.Map // uuid.UUID -> *GroupVoiceCallSession

	// Event callbacks — optional.
	OnInviteReceived    func(session *GroupVoiceCallSession, fromUhid string)
	OnMemberJoined      func(session *GroupVoiceCallSession, uhid string)
	OnMemberLeft        func(session *GroupVoiceCallSession, uhid string)
	OnMemberKicked      func(session *GroupVoiceCallSession, uhid string)
	OnCallEnded         func(session *GroupVoiceCallSession)
	OnKeyRotated        func(session *GroupVoiceCallSession)
	OnFrameReceived     func(session *GroupVoiceCallSession, frame *GroupVoiceFrame)
}

// GroupVoiceFrame is the decoded form of a binary group voice payload.
type GroupVoiceFrame struct {
	CallID         uuid.UUID
	Sequence       uint32
	TimestampMs    int64
	IsSilence      bool
	KeyGeneration  uint32
	EncodedPayload []byte
}

// NewGroupVoiceCallService constructs a GroupVoiceCallService.
func NewGroupVoiceCallService(sender routing.MeshSender) *GroupVoiceCallService {
	if sender == nil {
		panic("voice: group sender must not be nil")
	}
	return &GroupVoiceCallService{
		sender:    sender,
		localUhid: sender.LocalUhid(),
	}
}

// Invite sends group call invitations to the listed member UHIDs. The caller
// becomes host. Returns the new call ID.
func (s *GroupVoiceCallService) Invite(ctx context.Context, callID uuid.UUID, memberUhids []string) error {
	if len(memberUhids) == 0 {
		return errors.New("voice: memberUhids must not be empty")
	}
	if int32(len(memberUhids)) > constants.MaxGroupVoiceMembers {
		return fmt.Errorf("voice: exceeds MaxGroupVoiceMembers (%d)", constants.MaxGroupVoiceMembers)
	}

	session := s.getOrCreateSession(callID, s.localUhid)
	session.mu.Lock()
	for _, m := range memberUhids {
		session.Members[m] = struct{}{}
	}
	session.mu.Unlock()

	msg := GroupVoiceSignalingMessage{
		Kind:         "invite",
		CallID:       callID.String(),
		FromUhid:     s.localUhid,
		InvitedUhids: memberUhids,
	}
	for _, m := range memberUhids {
		msg.ToUhid = m
		if err := s.sendGroupSignaling(ctx, m, msg); err != nil {
			return err
		}
	}
	return nil
}

// Join sends a "join" signal to the host / all current members.
func (s *GroupVoiceCallService) Join(ctx context.Context, callID uuid.UUID) error {
	session, err := s.getGroupSession(callID)
	if err != nil {
		return err
	}
	session.mu.Lock()
	session.State = VoiceCallStateActive
	members := s.memberSnapshot(session)
	session.mu.Unlock()

	msg := GroupVoiceSignalingMessage{
		Kind:     "join",
		CallID:   callID.String(),
		FromUhid: s.localUhid,
	}
	return s.broadcastGroupSignaling(ctx, members, msg)
}

// Leave exits a group call and notifies the remaining members.
func (s *GroupVoiceCallService) Leave(ctx context.Context, callID uuid.UUID) error {
	session, err := s.getGroupSession(callID)
	if err != nil {
		return err
	}
	session.mu.Lock()
	session.State = VoiceCallStateEnded
	delete(session.Members, s.localUhid)
	members := s.memberSnapshot(session)
	session.mu.Unlock()

	msg := GroupVoiceSignalingMessage{
		Kind:     "leave",
		CallID:   callID.String(),
		FromUhid: s.localUhid,
	}
	s.calls.Delete(callID)
	return s.broadcastGroupSignaling(ctx, members, msg)
}

// Kick removes a member from the call (host-only by convention — enforcement is
// application-layer; the service sends the signal unconditionally).
func (s *GroupVoiceCallService) Kick(ctx context.Context, callID uuid.UUID, targetUhid string) error {
	if targetUhid == "" {
		return errors.New("voice: targetUhid must not be empty")
	}
	session, err := s.getGroupSession(callID)
	if err != nil {
		return err
	}
	session.mu.Lock()
	delete(session.Members, targetUhid)
	members := s.memberSnapshot(session)
	session.mu.Unlock()

	msg := GroupVoiceSignalingMessage{
		Kind:       "kick",
		CallID:     callID.String(),
		FromUhid:   s.localUhid,
		ToUhid:     targetUhid,
		KickedUhid: targetUhid,
	}
	// Notify the kicked member directly.
	if err := s.sendGroupSignaling(ctx, targetUhid, msg); err != nil {
		return err
	}
	// Notify remaining members.
	return s.broadcastGroupSignaling(ctx, members, msg)
}

// SendFrame encodes and transmits a group voice frame to all current members.
//
// Binary layout (GroupVoiceFrame):
//
//	[16] CallId         (UUID, big-endian)
//	[4]  Sequence       (uint32, little-endian)
//	[8]  TimestampMs    (int64, little-endian)
//	[1]  IsSilence      (0 or 1)
//	[4]  KeyGeneration  (uint32, little-endian)
//	[N]  EncodedPayload
func (s *GroupVoiceCallService) SendFrame(ctx context.Context, callID uuid.UUID, audio []byte, isSilence bool, keyGeneration uint32) error {
	session, err := s.getGroupSession(callID)
	if err != nil {
		return err
	}
	session.mu.Lock()
	session.Sequence++
	seq := session.Sequence
	members := s.memberSnapshot(session)
	session.mu.Unlock()

	payload, err := marshalGroupVoiceFrame(callID, seq, time.Now().UnixMilli(), isSilence, keyGeneration, audio)
	if err != nil {
		return err
	}

	for _, m := range members {
		if m == s.localUhid {
			continue
		}
		pkt := protocol.NewMeshPacket()
		pkt.Type = protocol.VoiceCall
		pkt.SourceUhid = s.localUhid
		pkt.DestinationUhid = m
		pkt.Ttl = constants.DefaultTtl
		pkt.Priority = 100
		pkt.Payload = payload
		_, _ = s.sender.Send(ctx, pkt, m)
	}
	return nil
}

// HandlePacket processes inbound group voice signaling or frame packets.
func (s *GroupVoiceCallService) HandlePacket(ctx context.Context, packet *protocol.MeshPacket) error {
	if packet == nil {
		return errors.New("voice: group packet must not be nil")
	}
	switch packet.Type {
	case protocol.VoiceSignaling:
		return s.handleGroupSignaling(ctx, packet)
	case protocol.VoiceCall:
		return s.handleGroupFrame(packet)
	default:
		return fmt.Errorf("voice: unexpected packet type %s", packet.Type)
	}
}

// ──────────────────────────────────────────────
// Internal helpers
// ──────────────────────────────────────────────

func (s *GroupVoiceCallService) handleGroupSignaling(_ context.Context, packet *protocol.MeshPacket) error {
	var msg GroupVoiceSignalingMessage
	if err := json.Unmarshal(packet.Payload, &msg); err != nil {
		return fmt.Errorf("voice: unmarshal group signaling: %w", err)
	}
	callID, err := uuid.Parse(msg.CallID)
	if err != nil {
		return fmt.Errorf("voice: invalid call_id %q: %w", msg.CallID, err)
	}

	switch msg.Kind {
	case "invite":
		session := s.getOrCreateSession(callID, packet.SourceUhid)
		session.mu.Lock()
		for _, m := range msg.InvitedUhids {
			session.Members[m] = struct{}{}
		}
		session.mu.Unlock()
		if cb := s.OnInviteReceived; cb != nil {
			cb(session, packet.SourceUhid)
		}

	case "join":
		if session, err := s.getGroupSession(callID); err == nil {
			session.mu.Lock()
			session.Members[packet.SourceUhid] = struct{}{}
			session.mu.Unlock()
			if cb := s.OnMemberJoined; cb != nil {
				cb(session, packet.SourceUhid)
			}
		}

	case "leave":
		if session, err := s.getGroupSession(callID); err == nil {
			session.mu.Lock()
			delete(session.Members, packet.SourceUhid)
			session.mu.Unlock()
			if cb := s.OnMemberLeft; cb != nil {
				cb(session, packet.SourceUhid)
			}
		}

	case "kick":
		if session, err := s.getGroupSession(callID); err == nil {
			session.mu.Lock()
			delete(session.Members, msg.KickedUhid)
			session.mu.Unlock()
			if cb := s.OnMemberKicked; cb != nil {
				cb(session, msg.KickedUhid)
			}
			if msg.KickedUhid == s.localUhid {
				session.mu.Lock()
				session.State = VoiceCallStateEnded
				session.mu.Unlock()
				s.calls.Delete(callID)
			}
		}

	case "end":
		if session, err := s.getGroupSession(callID); err == nil {
			session.mu.Lock()
			session.State = VoiceCallStateEnded
			session.mu.Unlock()
			if cb := s.OnCallEnded; cb != nil {
				cb(session)
			}
			s.calls.Delete(callID)
		}

	case "key_rotation":
		if session, err := s.getGroupSession(callID); err == nil {
			session.mu.Lock()
			session.KeyGeneration = msg.KeyGeneration
			session.mu.Unlock()
			if cb := s.OnKeyRotated; cb != nil {
				cb(session)
			}
		}
	}
	return nil
}

func (s *GroupVoiceCallService) handleGroupFrame(packet *protocol.MeshPacket) error {
	frame, err := unmarshalGroupVoiceFrame(packet.Payload)
	if err != nil {
		return fmt.Errorf("voice: unmarshal group frame: %w", err)
	}
	session, err := s.getGroupSession(frame.CallID)
	if err != nil {
		return nil // unknown call — drop
	}
	if cb := s.OnFrameReceived; cb != nil {
		cb(session, frame)
	}
	return nil
}

func (s *GroupVoiceCallService) getGroupSession(callID uuid.UUID) (*GroupVoiceCallSession, error) {
	v, ok := s.calls.Load(callID)
	if !ok {
		return nil, fmt.Errorf("voice: unknown group call %s", callID)
	}
	return v.(*GroupVoiceCallSession), nil
}

func (s *GroupVoiceCallService) getOrCreateSession(callID uuid.UUID, hostUhid string) *GroupVoiceCallSession {
	v, _ := s.calls.LoadOrStore(callID, &GroupVoiceCallSession{
		CallID:    callID,
		HostUhid:  hostUhid,
		Members:   make(map[string]struct{}),
		State:     VoiceCallStateOffering,
		CreatedAt: time.Now(),
	})
	return v.(*GroupVoiceCallSession)
}

func (s *GroupVoiceCallService) memberSnapshot(session *GroupVoiceCallSession) []string {
	out := make([]string, 0, len(session.Members))
	for m := range session.Members {
		out = append(out, m)
	}
	return out
}

func (s *GroupVoiceCallService) sendGroupSignaling(ctx context.Context, toUhid string, msg GroupVoiceSignalingMessage) error {
	body, err := json.Marshal(msg)
	if err != nil {
		return fmt.Errorf("voice: marshal group signaling: %w", err)
	}
	pkt := protocol.NewMeshPacket()
	pkt.Type = protocol.VoiceSignaling
	pkt.SourceUhid = s.localUhid
	pkt.DestinationUhid = toUhid
	pkt.Ttl = constants.DefaultTtl
	pkt.Payload = body
	_, err = s.sender.Send(ctx, pkt, toUhid)
	return err
}

func (s *GroupVoiceCallService) broadcastGroupSignaling(ctx context.Context, members []string, msg GroupVoiceSignalingMessage) error {
	for _, m := range members {
		if m == s.localUhid {
			continue
		}
		msg.ToUhid = m
		if err := s.sendGroupSignaling(ctx, m, msg); err != nil {
			return err
		}
	}
	return nil
}

// ──────────────────────────────────────────────
// Binary serialisation helpers
// ──────────────────────────────────────────────

// marshalGroupVoiceFrame encodes a GroupVoiceFrame into its canonical binary format.
func marshalGroupVoiceFrame(callID uuid.UUID, seq uint32, tsMs int64, isSilence bool, keyGen uint32, payload []byte) ([]byte, error) {
	var buf bytes.Buffer
	buf.Write(callID[:])
	if err := binary.Write(&buf, binary.LittleEndian, seq); err != nil {
		return nil, err
	}
	if err := binary.Write(&buf, binary.LittleEndian, tsMs); err != nil {
		return nil, err
	}
	if isSilence {
		buf.WriteByte(1)
	} else {
		buf.WriteByte(0)
	}
	if err := binary.Write(&buf, binary.LittleEndian, keyGen); err != nil {
		return nil, err
	}
	buf.Write(payload)
	return buf.Bytes(), nil
}

// unmarshalGroupVoiceFrame decodes a binary GroupVoiceFrame payload.
func unmarshalGroupVoiceFrame(data []byte) (*GroupVoiceFrame, error) {
	const fixedSize = 16 + 4 + 8 + 1 + 4
	if len(data) < fixedSize {
		return nil, fmt.Errorf("voice: group frame too short: %d bytes", len(data))
	}
	r := bytes.NewReader(data)

	var callIDBytes [16]byte
	if _, err := r.Read(callIDBytes[:]); err != nil {
		return nil, err
	}
	callID, err := uuid.FromBytes(callIDBytes[:])
	if err != nil {
		return nil, err
	}

	var seq uint32
	if err := binary.Read(r, binary.LittleEndian, &seq); err != nil {
		return nil, err
	}

	var tsMs int64
	if err := binary.Read(r, binary.LittleEndian, &tsMs); err != nil {
		return nil, err
	}

	silenceByte, err := r.ReadByte()
	if err != nil {
		return nil, err
	}

	var keyGen uint32
	if err := binary.Read(r, binary.LittleEndian, &keyGen); err != nil {
		return nil, err
	}

	encoded := make([]byte, r.Len())
	if _, err := r.Read(encoded); err != nil && r.Len() > 0 {
		return nil, err
	}

	return &GroupVoiceFrame{
		CallID:         callID,
		Sequence:       seq,
		TimestampMs:    tsMs,
		IsSilence:      silenceByte != 0,
		KeyGeneration:  keyGen,
		EncodedPayload: encoded,
	}, nil
}

File: rust/src/voice/group_service.rs
// SPDX-License-Identifier: MIT

//! Group voice call service: multi-party signaling + binary frame delivery.

use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;
use tokio::sync::broadcast;

use crate::constants::DEFAULT_TTL;
use crate::protocol::{MeshPacket, PacketType};
use crate::routing::sender::MeshSender;
use crate::routing::RoutingService;

// ── Signaling wire type ────────────────────────────────────────────────────────

/// JSON signaling for group voice. Cross-language canonical wire format.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct GroupVoiceSignalingMessage {
    pub kind: String,
    pub call_id: Uuid,
    pub from_uhid: String,
    pub to_uhid: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub invited_uhids: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kicked_uhid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub key_generation: Option<u32>,
}

// ── Group call state ───────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct GroupCallEntry {
    pub call_id: Uuid,
    pub host_uhid: String,
    pub members: HashSet<String>,
    pub active: bool,
}

// ── Inbound frame event ────────────────────────────────────────────────────────

/// Capacity of the inbound-frame broadcast channel; lagging subscribers drop oldest.
const FRAME_RECEIVED_CHANNEL_CAPACITY: usize = 256;

/// An inbound group-voice frame decoded from a received VOICE_CALL packet, delivered
/// to subscribers of [`GroupVoiceCallService::subscribe_frames`].
#[derive(Debug, Clone)]
pub struct GroupVoiceFrameReceivedEvent {
    pub call_id: Uuid,
    pub sender_uhid: String,
    pub seq: u32,
    pub timestamp_ms: i64,
    pub is_silence: bool,
    pub key_generation: u32,
    pub audio: Vec<u8>,
}

// ── Service ───────────────────────────────────────────────────────────────────

pub struct GroupVoiceCallService {
    sender: Arc<dyn MeshSender>,
    routing: Arc<RoutingService>,
    calls: Mutex<HashMap<Uuid, GroupCallEntry>>,
    frame_received_tx: broadcast::Sender<GroupVoiceFrameReceivedEvent>,
}

impl GroupVoiceCallService {
    pub fn new(sender: Arc<dyn MeshSender>, routing: Arc<RoutingService>) -> Self {
        let (frame_received_tx, _) = broadcast::channel(FRAME_RECEIVED_CHANNEL_CAPACITY);
        Self {
            sender,
            routing,
            calls: Mutex::new(HashMap::new()),
            frame_received_tx,
        }
    }

    /// Subscribe to inbound group-voice frames decoded from received VOICE_CALL packets.
    pub fn subscribe_frames(&self) -> broadcast::Receiver<GroupVoiceFrameReceivedEvent> {
        self.frame_received_tx.subscribe()
    }

    /// Invite a set of members to a (new or existing) group call.
    pub async fn invite(
        &self,
        call_id: Uuid,
        member_uhids: &[String],
    ) -> Result<(), String> {
        if member_uhids.is_empty() {
            return Err("member_uhids is empty".into());
        }
        let local = self.sender.local_uhid();
        {
            let mut calls = self.calls.lock().unwrap();
            let entry = calls.entry(call_id).or_insert_with(|| GroupCallEntry {
                call_id,
                host_uhid: local.clone(),
                members: HashSet::new(),
                active: true,
            });
            for m in member_uhids {
                entry.members.insert(m.clone());
            }
        }
        for uhid in member_uhids {
            let sig = GroupVoiceSignalingMessage {
                kind: "invite".into(),
                call_id,
                from_uhid: local.clone(),
                to_uhid: uhid.clone(),
                invited_uhids: Some(member_uhids.to_vec()),
                kicked_uhid: None,
                key_generation: None,
            };
            self.send_signaling(uhid, &sig).await;
        }
        Ok(())
    }

    /// Join a group call you were invited to.
    pub async fn join(&self, call_id: Uuid) -> Result<(), String> {
        let (local, members) = {
            let mut calls = self.calls.lock().unwrap();
            let entry = calls
                .get_mut(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?;
            entry.members.insert(self.sender.local_uhid());
            (self.sender.local_uhid(), entry.members.clone())
        };
        for uhid in &members {
            if *uhid == local {
                continue;
            }
            let sig = GroupVoiceSignalingMessage {
                kind: "join".into(),
                call_id,
                from_uhid: local.clone(),
                to_uhid: uhid.clone(),
                invited_uhids: None,
                kicked_uhid: None,
                key_generation: None,
            };
            self.send_signaling(uhid, &sig).await;
        }
        Ok(())
    }

    /// Leave a group call.
    pub async fn leave(&self, call_id: Uuid) -> Result<(), String> {
        let (local, members) = {
            let mut calls = self.calls.lock().unwrap();
            let entry = calls
                .get_mut(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?;
            entry.members.remove(&self.sender.local_uhid());
            (self.sender.local_uhid(), entry.members.clone())
        };
        for uhid in &members {
            let sig = GroupVoiceSignalingMessage {
                kind: "leave".into(),
                call_id,
                from_uhid: local.clone(),
                to_uhid: uhid.clone(),
                invited_uhids: None,
                kicked_uhid: None,
                key_generation: None,
            };
            self.send_signaling(uhid, &sig).await;
        }
        Ok(())
    }

    /// Kick a member from the call. Only the host may do this; enforced
    /// locally — the protocol relies on the host propagating the kick signal.
    pub async fn kick(
        &self,
        call_id: Uuid,
        target_uhid: &str,
    ) -> Result<(), String> {
        let (local, members) = {
            let mut calls = self.calls.lock().unwrap();
            let entry = calls
                .get_mut(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?;
            if entry.host_uhid != self.sender.local_uhid() {
                return Err("only the host may kick members".into());
            }
            entry.members.remove(target_uhid);
            (self.sender.local_uhid(), entry.members.clone())
        };
        for uhid in &members {
            let sig = GroupVoiceSignalingMessage {
                kind: "kick".into(),
                call_id,
                from_uhid: local.clone(),
                to_uhid: uhid.clone(),
                invited_uhids: None,
                kicked_uhid: Some(target_uhid.to_string()),
                key_generation: None,
            };
            self.send_signaling(uhid, &sig).await;
        }
        Ok(())
    }

    /// Send an encoded audio frame to all current group members.
    ///
    /// Binary payload wire format:
    ///   [16] CallId (UUID RFC4122 big-endian)
    ///   [4]  Sequence (u32 little-endian)
    ///   [8]  TimestampMs (i64 little-endian)
    ///   [1]  IsSilence (0 or 1)
    ///   [4]  KeyGeneration (u32 little-endian)
    ///   [N]  EncodedPayload
    pub async fn send_frame(
        &self,
        call_id: Uuid,
        encoded_audio: &[u8],
        is_silence: bool,
        key_generation: u32,
    ) -> Result<(), String> {
        let members = {
            let calls = self.calls.lock().unwrap();
            let entry = calls
                .get(&call_id)
                .ok_or_else(|| format!("unknown call_id {}", call_id))?;
            if !entry.active {
                return Err(format!("call {} is not active", call_id));
            }
            entry.members.clone()
        };

        let payload = build_group_voice_frame(call_id, encoded_audio, is_silence, key_generation);
        let local = self.sender.local_uhid();

        for uhid in &members {
            if *uhid == local {
                continue;
            }
            let mut pkt = MeshPacket::new(PacketType::VoiceCall, local.clone());
            pkt.destination_uhid = uhid.clone();
            pkt.ttl = DEFAULT_TTL;
            pkt.priority = 64;
            pkt.payload = payload.clone();
            self.sender.send(&pkt, uhid).await;
        }
        Ok(())
    }

    /// Route an inbound packet to the right handler.
    pub async fn handle_packet(&self, packet: &MeshPacket) -> Result<(), String> {
        match packet.packet_type {
            PacketType::VoiceSignaling => self.handle_signaling(packet).await,
            PacketType::VoiceCall => self.handle_frame(packet).await,
            _ => Ok(()),
        }
    }

    /// Decode an inbound group-voice frame and surface it to application subscribers.
    /// Layout (build_group_voice_frame): [16] call_id | [4] seq LE | [8] ts LE |
    /// [1] silence | [4] key_generation LE | [N] audio.
    async fn handle_frame(&self, packet: &MeshPacket) -> Result<(), String> {
        let p = &packet.payload;
        if p.len() < 33 {
            return Err("group voice frame too short".to_string());
        }
        let call_id = Uuid::from_slice(&p[0..16]).map_err(|e| e.to_string())?;
        let seq = u32::from_le_bytes(p[16..20].try_into().unwrap());
        let timestamp_ms = i64::from_le_bytes(p[20..28].try_into().unwrap());
        let is_silence = p[28] != 0;
        let key_generation = u32::from_le_bytes(p[29..33].try_into().unwrap());
        let audio = p[33..].to_vec();
        let _ = self.frame_received_tx.send(GroupVoiceFrameReceivedEvent {
            call_id,
            sender_uhid: packet.source_uhid.clone(),
            seq,
            timestamp_ms,
            is_silence,
            key_generation,
            audio,
        });
        Ok(())
    }

    // ── private ────────────────────────────────────────────────────────────────

    async fn send_signaling(&self, to_uhid: &str, sig: &GroupVoiceSignalingMessage) {
        let body = match serde_json::to_vec(sig) {
            Ok(b) => b,
            Err(_) => return,
        };
        let route = self.routing.find_route(to_uhid).await;
        let next_hop = route
            .map(|r| r.next_hop_uhid)
            .unwrap_or_else(|| to_uhid.to_string());

        let mut pkt = MeshPacket::new(PacketType::VoiceSignaling, self.sender.local_uhid());
        pkt.destination_uhid = to_uhid.to_string();
        pkt.ttl = DEFAULT_TTL;
        pkt.priority = 32;
        pkt.payload = body;
        self.sender.send(&pkt, &next_hop).await;
    }

    async fn handle_signaling(&self, packet: &MeshPacket) -> Result<(), String> {
        let sig: GroupVoiceSignalingMessage = serde_json::from_slice(&packet.payload)
            .map_err(|e| format!("bad group signaling payload: {}", e))?;

        let local = self.sender.local_uhid();
        let mut calls = self.calls.lock().unwrap();
        match sig.kind.as_str() {
            "invite" => {
                let entry = calls.entry(sig.call_id).or_insert_with(|| GroupCallEntry {
                    call_id: sig.call_id,
                    host_uhid: packet.source_uhid.clone(),
                    members: HashSet::new(),
                    active: true,
                });
                if let Some(invited) = &sig.invited_uhids {
                    for m in invited {
                        entry.members.insert(m.clone());
                    }
                }
            }
            "join" => {
                if let Some(entry) = calls.get_mut(&sig.call_id) {
                    entry.members.insert(packet.source_uhid.clone());
                }
            }
            "leave" => {
                if let Some(entry) = calls.get_mut(&sig.call_id) {
                    entry.members.remove(&packet.source_uhid);
                }
            }
            "kick" => {
                if let Some(entry) = calls.get_mut(&sig.call_id) {
                    if let Some(kicked) = &sig.kicked_uhid {
                        entry.members.remove(kicked);
                        if kicked == &local {
                            entry.active = false;
                        }
                    }
                }
            }
            "end" => {
                if let Some(entry) = calls.get_mut(&sig.call_id) {
                    entry.active = false;
                }
            }
            _ => {}
        }
        Ok(())
    }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

static GROUP_FRAME_SEQ: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);

fn build_group_voice_frame(
    call_id: Uuid,
    encoded_audio: &[u8],
    is_silence: bool,
    key_generation: u32,
) -> Vec<u8> {
    let seq = GROUP_FRAME_SEQ.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
    let ts = unix_millis();
    let mut buf = Vec::with_capacity(33 + encoded_audio.len());
    buf.extend_from_slice(call_id.as_bytes()); // [16]
    buf.extend_from_slice(&seq.to_le_bytes()); // [4]
    buf.extend_from_slice(&ts.to_le_bytes()); // [8]
    buf.push(if is_silence { 1 } else { 0 }); // [1]
    buf.extend_from_slice(&key_generation.to_le_bytes()); // [4]
    buf.extend_from_slice(encoded_audio); // [N]
    buf
}

fn unix_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}
